flang/docs/ArrayRepacking.md

```diff
@@ -0,0 +1,498 @@
+
+* `MemFree` effect on either `AutomaticAllocationScopeResource` or `DefaultResource` depending on `stack` attribute.
+* If there is no `no_copy`:
+  * `MemRead` effect on unknown value to indicate potential read from the temporary array.
+  * `MemWrite` effect on unknown value to indicate potential write into the original array.
+
+### New attribute interface
+
+The `TempCopyIsSafe` attribute interface provides means to generate programming model specific predicates saying whether repacking is safe or not at the point where it needs to be done. For example the OpenMP MLIR dialect may provide an attribute implementing this interface to generate a runtime check at the point of packing array `x` inside subroutine `repacking`. A conservative implementation might look like this:
+
+```C
+  repacking_is_safe = omp_get_num_team() == 1 && omp_get_num_threads() == 1;
+```
+
+The attribute interface will be used during lowering of `fir.pack_array` to generate the predicate and make the packing conditional and safe. This will allow applying repacking correctly in programs compiled with `-fopenmp`, and get the benefits of repacking in the serial parts of those programs.
+
+Similarly, the OpenACC MLIR dialect may provide such an attribute to check if a device copy has been created for an array that is about to be packed, and prevent the repacking for cases like this:
+
+```Fortran
+program main
+  real :: x(2,10)
+  x = 0.0
+  !$acc enter data copyin(x)
+  do i=1,2
+     call repacking(x(i,:))
+  end do
+  !$acc exit data copyout(x)
+  print *, x
+contains
+  subroutine repacking(x)
+    real :: x(:)
+    !$acc parallel loop present(x)
+    do j=1,10
+       x(j) = x(j) + 1.0
+    end do
+  end subroutine repacking
+end program main
+```
+
+Note that it may not be possible to detect all the cases where repacking breaks the original OpenACC/OpenMP program behavior. In the following case, it is unclear how to synchronize the repacking in both routines (especially, when the OpenACC directives are not lexically visible at the point of repacking):
+
+```Fortran
+program main
+  real :: x(2,10)
+  x = 0.0
+  call repacking_in(x(1,:))
+  call repacking_out(x(1,:))
+  print *, x
+contains
+  subroutine repacking_in(x)
+    real :: x(:)
+    !$acc enter data copyin(x)
+  end subroutine repacking_in
+  subroutine repacking_out(x)
+    real :: x(:)
+    !$acc exit data copyout(x)
+  end subroutine repacking_out
+end program main
+```
+
+So the most conservative implementation of the predicate generator may be to always produce `false` value. Flang lowering may attach any number of such attributes to `fir.pack_array` depending on the compilation context and options.
+
+[TBD] define `TempCopyIsSafe` attribute interface so that OpenACC/OpenMP dialects can provide their specific attributes, which can be used to generate static/runtime checks for safety of the temporary copy in particular context.
```

Comment:
As you pointed out earlier, what do we do about cases where the runtime is not going to be linked (or at least we don't know that it will be linked from the context of compiling this translation unit)? I guess this should just be best effort?

Reply:
Right, we can only rely on OpenACC/OpenMP runtime support if the libraries are eventually linked.  So we can only generate the extra stuff when OpenACC/OpenMP is enabled consistently for all compilation modules.

I tried to explain that in the `Correctness` requirements section with `When array repacking is enabled, Flang should guarantee correct program behavior when OpenACC/OpenMP features are explicitly enabled during the compilation.`